cartridge/upload.lua

```diff
@@ -0,0 +1,322 @@
+        'cartridge-upload.' .. upload_id
+    )
+end
+
+-- The upload starts by creating a shared resource - a directory named
+-- after the `upload_id`. The instances who can create it before
+-- others are the transmitters. The others wait patiently till the
+-- `upload_finish()`.
+local function upload_begin(upload_id)
+    checks('string')
+
+    vars.upload_fibers[upload_id] = fiber.self()
+
+    local upload_path = get_upload_path(upload_id)
+    local ok, err = utils.mktree(fio.dirname(upload_path))
+    fiber.testcancel()
+    if ok == nil then
+        vars.upload_fibers[upload_id] = nil
+        return nil, err
+    end
+
+    local ok = fio.mkdir(upload_path)
+    local _errno = errno()
+    fiber.testcancel()
+
+    vars.upload_fibers[upload_id] = nil
+
+    if ok then
+        return true
+    elseif fio.path.is_dir(upload_path) then
+        return false
+    else
+        return nil, UploadError:new(
+            'Error creating directory %q: %s',
+            upload_path, errno.strerror(_errno)
+        )
+    end
+end
+
+-- The communication only goes with transmitters - the instances who
+-- replied `upload_begin() == true` i.e affirmed they own the shared
+-- resource. The other instances will read the data from that resource
+-- during the `upload_finish()` step.
+local function upload_transmit(upload_id, payload)
+    checks('string', 'string')
+
+    vars.upload_fibers[upload_id] = fiber.self()
+
+    local upload_path = get_upload_path(upload_id)
+    local payload_path = fio.pathjoin(upload_path, 'payload')
+    local ok, err = utils.file_write(payload_path, payload)
+    fiber.testcancel()
+    if not ok then
+        vars.upload_fibers[upload_id] = nil
+        return nil, err
+    end
+
+    vars.upload_fibers[upload_id] = nil
+    return true
+end
+
+-- The uploaded data is read from the shared resource on every instance,
+-- not only on the transmitter.
+local function upload_finish(upload_id)
+    checks('string')
+
+    vars.upload_fibers[upload_id] = fiber.self()
+
+    local upload_path = get_upload_path(upload_id)
+    local payload_path = fio.pathjoin(upload_path, 'payload')
+    local payload, err = utils.file_read(payload_path)
+
+    vars.upload_fibers[upload_id] = nil
+
+    fiber.testcancel()
+    if err then
+        return nil, err
+    end
+
+    local ok, data = pcall(msgpack.decode, payload)
+    if not ok then
+        return nil, UploadError:new(data)
+    end
+    vars.inbox[upload_id] = data
+    return true
+end
+
+-- Remove shared resources. Be idempotent.
+local function upload_cleanup(upload_id)
+    checks('string')
+    if vars.upload_fibers[upload_id] ~= nil then
+        pcall(fiber.cancel, vars.upload_fibers[upload_id])
+        vars.upload_fibers[upload_id] = nil
+    end
+
+    local upload_path = get_upload_path(upload_id)
+    local random_path = utils.randomize_path(upload_path)
+
+    local ok = fio.rename(upload_path, random_path)
+    local _errno = errno()
```

Comment:
It's redundant action to call errno here. It's not needed if ok is true.

Reply:
The main danger of errno is that it may be overridden by some other consequent fio operation. I've changed this place, but in another way.

Reply:
```lua
local _errno
local ok, err = fio.some_operation()
if not ok then
   _errno = errno()
end
```

No redefinition, to excess actions.